src/user-service.ts

```diff
@@ -52,18 +52,40 @@ export class UserService implements UserServiceInterface {
 
   /** @inheritdoc */
   async getLoggedInUser(): Promise<Result<UserInterface, UserServiceError>> {
+    // if the user doesn't have IA cookies, just return a `userNotLoggedIn` error
     const hasCookies = await this.hasArchiveOrgLoggedInCookies();
     if (!hasCookies)
       return {
         error: new UserServiceError(UserServiceErrorType.userNotLoggedIn),
       };
 
+    // check for cached user
     const persistedUser = await this.getPersistedUser();
     if (persistedUser) {
       const user = User.fromUserResponse(persistedUser);
       return { success: user };
     }
 
+    // if another fetch is in progress, chain this request to it
+    if (this.fetchPromise) {
+      this.fetchPromise = this.fetchPromise.then(response => {
+        return response;
```

Comment:
QUESTION: I'm probably just not understanding the use of promises here correctly, but how does the user info get wrapped into the promise chain?

Reply:
This is creating a new promise chained to the previous promise and returning that new chain to the new requestor. The very first request that comes through (a few lines down from here) assigns the first `fetchPromise` to the "actual" fetching code. 

Then a new request comes in and since `fetchPromise` already exists, it takes that promise, attaches the `then()` to it and just returns that initial response verbatim. This can keep happening until you have a chain of promises like:
```
firstPromise.then(response => secondPromise.then(response => thirdPromise.then(...)))
```
So when that very first promise gets resolved (the one doing the actual fetch work), it will send the response down the chain to all requestors in the chain.